### ISSUE TITLE
[spike] Emulator needs massive improvements

### ISSUE THREAD

**Issue by Cee123** [libretro/virtualjaguar-libretro#38]
Ok, so I tested out a bunch of Jaguar games and noticed a lot of them are broken or have issues. And decided to make a list of which ones have issues. This emulator seems to need a lot of work.

Air Cars
Half the title screen is cut off
![Air Cars](https://user-images.githubusercontent.com/43462099/60379176-f0436f80-9a60-11e9-8a99-1ecda001fdf1.png)

Atari Karts
Weird bar down the side of the screen during the game
![Atari Karts](https://user-images.githubusercontent.com/43462099/60379177-fc2f3180-9a60-11e9-96b4-b3005c76b7b1.png)

Battlesphere Gold
Only loads if Bios is enabled. But menus are glitched - black screen after menu, game doesn't work at all
![Battlesphere](https://user-images.githubusercontent.com/43462099/60379180-03eed600-9a61-11e9-9df3-c6e212cca315.png)

Club Drive
Half the title screen is cut off and textures are missing.
![Club Drive](https://user-images.githubusercontent.com/43462099/60379181-09e4b700-9a61-11e9-9faa-04c59ca3cb52.png)
![Club Drive 2](https://user-images.githubusercontent.com/43462099/60379182-0e10d480-9a61-11e9-9475-2be106412754.png)

Cybermorph
Missing/glitched textures
![Cybermorph](https://user-images.githubusercontent.com/43462099/60379184-123cf200-9a61-11e9-87df-8e863a5aa69a.png)

Doom
Weird bar down half of the screen during the game
![Doom](https://user-images.githubusercontent.com/43462099/60379185-1832d300-9a61-11e9-8cda-59fe858e85c8.png)

Fight for Life
Flickering graphics, mainly in the menu
![Fight for Life](https://user-images.githubusercontent.com/43462099/60379187-1ec14a80-9a61-11e9-888c-94f175a8197f.png)

Flashback - Quest for Identity
Game displays on the left side of the screen
![Flashback - Quest for Identity](https://user-images.githubusercontent.com/43462099/60379188-24b72b80-9a61-11e9-8272-bf57fada0765.png)
![Flashback Quest for Identity 2](https://user-images.githubusercontent.com/43462099/60379189-28e34900-9a61-11e9-952e-b3dd4f24d98b.png)

Hover Strike
Glitched and flickering graphics
![Hover Strike](https://user-images.githubusercontent.com/43462099/60379191-30a2ed80-9a61-11e9-91f7-ab5229c489f8.png)
![Hover Strike 2](https://user-images.githubusercontent.com/43462099/60379195-34367480-9a61-11e9-864c-5ac346222ccd.png)

Hyper Force
Black screen after title, doesn't work at all

I-War
Half the title screen is cut off, glitched and flickering graphics
![I-War](https://user-images.githubusercontent.com/43462099/60379198-3e587300-9a61-11e9-9006-bcf82e92b9d9.png)
![I-War 2](https://user-images.githubusercontent.com/43462099/60379199-42849080-9a61-11e9-8b8a-48911ac72f07.png)

Iron Soldier
Black screen after character selection, doesn't work at all

Iron Soldier 2
Title displays on the left side of the screen, black screen after title but music still plays, doesn't work at all
![Iron Soldier 2](https://user-images.githubusercontent.com/43462099/60379202-4e705280-9a61-11e9-8262-9092eafc9282.png)

Kasumi Ninja
Severe Graphics issues, flickering at the title screen, missing textures, black screen freezes
![Kasumi Ninja](https://user-images.githubusercontent.com/43462099/60379204-57f9ba80-9a61-11e9-9e2b-10740486292c.png)
![Kasumi Ninja2](https://user-images.githubusercontent.com/43462099/60379205-5af4ab00-9a61-11e9-9d90-2464707dd1f9.png)

Missile Command 3D
Some minor graphics glitches/flickering. Nothing too serious.
![Missile Command 3D](https://user-images.githubusercontent.com/43462099/60380150-47524000-9a73-11e9-9735-23ca3b05a59b.png)

NBA Jam - Tournament Edition
Some minor occasional screen jumping, nothing too serious.

Pinball Fantasies
Some of the introduction screens are cut off
![Pinball Fantasies](https://user-images.githubusercontent.com/43462099/60379963-854d6500-9a6f-11e9-9958-3804ee34548f.png)

Powerdrive Rally
Some of the introduction screens are cut off, game crashes upon starting a new game
![Powerdrive Rally](https://user-images.githubusercontent.com/43462099/60379966-8e3e3680-9a6f-11e9-96d8-adb82d06a4ed.png)

Raiden
Only works if the BIOS is enabled, otherwise black screen freezes

Ruiner Pinball
Black screen, doesn't work at all

Skyhammer
Works but freezes during game

Super Burnout
Some of the letters are cut off, black screen doesnt work at all when starting a new game
![Super Burnout](https://user-images.githubusercontent.com/43462099/60379967-9bf3bc00-9a6f-11e9-92c1-dc16a88f8c7d.png)

Supercross 3D
Displays on the left side of the screen, blurry textures
![Supercross 3D](https://user-images.githubusercontent.com/43462099/60379971-a4e48d80-9a6f-11e9-95b4-7c8cebd1beb8.png)

Syndicate
Works good, but there are pink patches in the writing
![Syndicate](https://user-images.githubusercontent.com/43462099/60379974-aada6e80-9a6f-11e9-9998-24a48950096a.png)

Tempest 2000
Works good, but there's some minor flickering occasionally

Towers II
Unplayable due to flickering/flashing

Trevor McFur in the Crescent Galaxy
Half the screen is cut off during the loading screen
![Trevor McFur in the Crescent Galaxy](https://user-images.githubusercontent.com/43462099/60379978-b7f75d80-9a6f-11e9-8781-5095bfeeef58.png)

Val D'isere Skiing and Snowboarding
Game displays on the left side of the screen
![Val D'isere Skiing and Snowboarding](https://user-images.githubusercontent.com/43462099/60379981-bfb70200-9a6f-11e9-8dc5-61b8eb27c134.png)

White Men Can't Jump
Black screen, doesn't work at all

Wolfenstein 3D
Black screen, doesn't work at all

Zero 5
Game works good, but some mild flickering

Zoop
Game crashes and doesn't work

**Comment by Cee123** [comment]
Bounty posted to fix issues with this core - $250
https://www.bountysource.com/issues/76219168-list-of-games-with-issues

**Comment by Chlorophytus** [comment]
Speaking of, I was going to teach myself how to do Atari Jaguar code a few days ago. I will try to help with this, maybe I can get to a point.

**Comment by Cee123** [comment]
Nice. Hoping someone can pick this up again, seems a bit broken in its current state.

**Comment by Tasosgemah** [comment]
Most of these issues (if not all) apply to the standalone emulator which hasn't been updated since 2014 IIRC. So they are not exclusive libretro core issues.

Also, the latest standalone version is 2.1.3 while the RetroArch core seems to be based on the older 2.1.0 version from 2013. Though, i don't think there's a huge difference in compatibility or if any new fixes exist for the core exclusively since then.

Oh and btw, there is a "DOOM Res hack" switch in the core options that fixes DOOM's screen but the emulation of the game is very slow anyway.

**Comment by Cee123** [comment]
You are right. The standalone emulator hasn't been updated since 2014. It has the same issues also. The only properly working Jaguar emulator is Irata Jaguar for Android. That, and Real3doplayer. But sadly, it's closed source and paid only. :( 

I reached out to the dev to see if he'd make his emulators open source or at least port them to other OSes but he said he's too busy.

Yeah I was aware of the Doom hack, seems to fix the problem for that gamr. But that's only one game.  And yeah unfortunately it runs too slow.

**Comment by Tasosgemah** [comment]
Irata Jaguar is based on the Russian Phoenix emulator for PC. It's the only good Jaguar emulator and yes, it's closed source. So no RetroArch fun with this one :(

**Comment by Chlorophytus** [comment]
I'll work on a Jaguar emulator called `jagger` but I will be doing it in a slow manner due to circumstances.
I will try to make it portable too. If I have to, I will use Arm NEON and x86 SSE (no AVX, Silvermont/Goldmont(+) SBCs will `SIGILL` the moment a >=256-bit vector operation is encountered), but I really just want use pure C.

I do have Gemini Lake hardware for those tight corner cases.

**Comment by Cee123** [comment]
That's awesome! You're a legend.

**Comment by Cee123** [comment]
I think it may be a few of these issues may be down to resolution. Just like the Room Res Hack seems to fix Doom. I don't know why they couldn't do similar for Atari Karts, Supercross Rally and Air Cars.

**Comment by Cee123** [comment]
A few of these issues don't seem to exist in the standalone virtual jaguar.

**Comment by Tasosgemah** [comment]
Well, the standalone is version 2.1.2 while the core is based on an older 2.1.0 version it seems.

**Comment by Cee123** [comment]
@Tasosgemah Really?? Damn.

**Comment by Cee123** [comment]
Might retract the bounty (we'll see). Seems that no one's interested.

**Comment by Cee123** [comment]
Does this emulator use Interpreter or Dynamic Recompiler? Seems like it runs on Interpreter, hence the speed issues. Maybe I should've opened a bounty for a Dynamic Recompiler.

**Comment by Tasosgemah** [comment]
I don't think the issue with this emulator is it's speed. There are major compatibility and accuracy issues. Everything runs at full speed otherwise on any low-end system.

**Comment by Cee123** [comment]
> 
> 
> I don't think the issue with this emulator is it's speed. There are major compatibility and accuracy issues. Everything runs at full speed otherwise on any low-end system.

I've tried it on a bunch of things. It runs slow all the time. On Raspberry Pi 3B, Odroid XU4 (which runs N64, 3DO, Dreamcast even most PSP full speed), Odroid N2 (which is even more powerful than the XU4), Raspberry Pi 4, my Android box, I've tried it on my HTC 10 phone. I can't ever seem to get it to run at a playable speed on anything except my PC. :(

**Comment by Cee123** [comment]
> 
> 
> I don't think the issue with this emulator is it's speed. There are major compatibility and accuracy issues. Everything runs at full speed otherwise on any low-end system.

Full speed on any low-end system? This core runs slow on everything I've tried (HTC Phone, Samsung Phone, Raspberry Pi 3B, Odroid XU4, Odroid N2, Raspberry Pi 4, Android Q Box+). Except for PC. And I mean some serious, serious lagging. But even on PC, some games are painfully slow or don't work at all. And some games run like they're in fast forward.

**Comment by Cee123** [comment]
Just noticed there's a Virtualjaguar RX available. I downloaded it and it seems to work pretty well so far. I wonder if they'd be willing to do a libretro port of it either now or in future.

**Comment by Tasosgemah** [comment]
> 
> 
> > I don't think the issue with this emulator is it's speed. There are major compatibility and accuracy issues. Everything runs at full speed otherwise on any low-end system.
> 
> Full speed on any low-end system? This core runs slow on everything I've tried (HTC Phone, Samsung Phone, Raspberry Pi 3B, Odroid XU4, Odroid N2, Raspberry Pi 4, Android Q Box+). Except for PC. And I mean some serious, serious lagging. But even on PC, some games are painfully slow or don't work at all. And some games run like they're in fast forward.

Ok maybe it doesn't run full speed on the lowest end systems. It does run full speed on a Core 2 Duo from 2006 so i figured it's fast enough. But still, it's not the speed that is the main issue here. Even on a fast PC at least half of the software doesn't run correctly or at all.

**Comment by Cee123** [comment]
> 
> 
> > > I don't think the issue with this emulator is it's speed. There are major compatibility and accuracy issues. Everything runs at full speed otherwise on any low-end system.
> > 
> > 
> > Full speed on any low-end system? This core runs slow on everything I've tried (HTC Phone, Samsung Phone, Raspberry Pi 3B, Odroid XU4, Odroid N2, Raspberry Pi 4, Android Q Box+). Except for PC. And I mean some serious, serious lagging. But even on PC, some games are painfully slow or don't work at all. And some games run like they're in fast forward.
> 
> Ok maybe it doesn't run full speed on the lowest end systems. It does run full speed on a Core 2 Duo from 2006 so i figured it's fast enough. But still, it's not the speed that is the main issue here. Even on a fast PC at least half of the software doesn't run correctly or at all.

Yeah I couldn't agree more. I'm surprised no one has taken up this project and tried to improve on it. We've seen massive improvements for Dreamcast, PSX, N64 and even 3DO. But Jaguar has continued to elude the emulation community.

I hope Virtualjaguar Rx ends up fixing some issues and that they also make a port for Libretro. That's seeming to be the only hope at the moment. Unless someone can unofficially and illegally port Irata Jaguar over since the original developer refuses.

**Comment by Cee123** [comment]
I've decided to retract the bounty on this. It's been well over a year now and nothing's happened. My bounty of $250 has been refunded. During that time, unfortunately I saw no contributions or anyone volunteering to take on the bounty. Hopefully we do some kind of continued development on a Jaguar emulator in the future but I'm not holding my breath anymore.

**Comment by Tasosgemah** [comment]
Yeah, virtual jaguar is at this state for more than a decade now so i don't think anyone cares about it anymore. There are better chances for some new emulator to pop up in the future or MAME surpassing it. Too bad because Jaguar emulation is one of the few gaps left for RetroArch.

**Comment by Cee123** [comment]
It's a real shame. Jaguar and also CD-i emulation are lacking. There is a Virtualjaguar Rx but it doesn't seem to have made many improvements over the original. And there is Phoenix Emulator but it's closed source and not only has the developer expressed total disinterest in making a retroarch port (dismissed it completely), he has also shown total disinterest in his own emulators and won't respond to messages. This whole thing is a bit of a disappointment.

**Comment by Chlorophytus** [comment]
I was working on an emulator that uses Musashi for the 68000 manager processor and probably some complex stuff elsewhere but I feel like I’ve taken on too many projects. CD-I emulation could be nice since those apparently have timekeeper modules that are savefile memory. I am not sure, but Jaguar is better to code since there’s docs out there. The documentation for what I think are the actual Atari ASICs used in the console are out, and the owners let it fly.

I will get some rest. I am interested.

Sent from my iPad

On Mar 26, 2020, at 9:59 PM, Cee123 <notifications@github.com> wrote:

﻿


It's a real shame. Jaguar and also CD-i emulation are lacking. There is a Virtualjaguar Rx but it doesn't seem to have made many improvements over the original. And there is Phoenix Emulator but it's closed source and not only has the developer expressed total disinterest in making a retroarch port (dismissed it completely), he has also shown total disinterest in his own emulators and won't respond to messages. This whole thing is a bit of a disappointment.

**Comment by Cee123** [comment]
Well I tested a bunch of games on the Phoenix Emulator if anyone is interested. Most of the issues aren't experienced using that emulator but there are a couple that are still the same. 

Air Cars - Displays properly
![Air Cars (World)@Sat April 11 2020 11-13-48-038](https://user-images.githubusercontent.com/43462099/79042860-c2b16680-7c2d-11ea-84a9-dcc11251700b.jpg)

Atari Karts - No annoying bar down the side of the screen
![Atari Karts (World)@Sat April 11 2020 11-14-35-622](https://user-images.githubusercontent.com/43462099/79042870-da88ea80-7c2d-11ea-84bb-b64cbc8484f1.jpg)

Battle Sphere Gold - The menus are still glitchy but in different parts.
![Battle Sphere Gold (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-23-02-427](https://user-images.githubusercontent.com/43462099/79042886-f9877c80-7c2d-11ea-88f0-c66624cba639.jpg)

Club Drive - Displays properly and doesn't crash
![Club Drive (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-23-43-661](https://user-images.githubusercontent.com/43462099/79042892-086e2f00-7c2e-11ea-8c2c-7d8c1e0ae99a.jpg)
![Club Drive (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-23-58-013](https://user-images.githubusercontent.com/43462099/79042894-0c01b600-7c2e-11ea-8cd4-caa97c312b09.jpg)

Cybermorph - No glitches
![Cybermorph (World) (Rev 2)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-25-34-557](https://user-images.githubusercontent.com/43462099/79042902-1fad1c80-7c2e-11ea-88f6-0d87b284a387.jpg)

Doom - No annoying bar on the side of the screen
![Doom (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-26-01-863](https://user-images.githubusercontent.com/43462099/79042907-2a67b180-7c2e-11ea-9371-e77a7fba6fe7.jpg)

Fight for Life - Now this is interesting. There is also a considerable amount of flickering/glitching with the Phoenix Emulator
![Fight for Life (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-26-59-740](https://user-images.githubusercontent.com/43462099/79042923-48351680-7c2e-11ea-9304-b86db4fbe324.jpg)

Flashback - The Quest for Identity - Displays properly
![Flashback - The Quest for Identity (World) (En,Fr)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-27-34-264](https://user-images.githubusercontent.com/43462099/79042934-5daa4080-7c2e-11ea-8848-46e0aa38c960.jpg)
![Flashback - The Quest for Identity (World) (En,Fr)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-28-00-797](https://user-images.githubusercontent.com/43462099/79042936-61d65e00-7c2e-11ea-8362-783cf74d536e.jpg)

Hover Strike - Displays properly, but interestingly it freezes during the start of a new game on the Phoenix Emulator. This doesn't happen on Virtual Jaguar
![Hover Strike (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-28-27-039](https://user-images.githubusercontent.com/43462099/79042953-80d4f000-7c2e-11ea-9c71-0d94533fdccb.jpg)
![Hover Strike (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-28-44-133](https://user-images.githubusercontent.com/43462099/79042954-84687700-7c2e-11ea-84f7-7a64f17b7531.jpg)
![Hover Strike (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-29-21-098](https://user-images.githubusercontent.com/43462099/79042956-88949480-7c2e-11ea-85b2-4a62a09519f2.jpg)

Hyperforce - Works! It's playable on Phoenix, but completely unplayable and not supported on Virtual Jaguar
![Hyper Force (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-30-07-815](https://user-images.githubusercontent.com/43462099/79042981-ab26ad80-7c2e-11ea-9939-7f868f827aa5.jpg)
![Hyper Force (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-30-27-756](https://user-images.githubusercontent.com/43462099/79042984-aeba3480-7c2e-11ea-9891-93dba24aa23c.jpg)

Iron Soldier - Interestingly, this game didn't work on Phoenix Emulator either. It just crashes to a black screen after character selection. But the music still plays.

Iron Soldier 2 - Title displays properly. Interestingly, this also doesn't work - crashes to a black screen upon the beginning of a new game.
![Iron Soldier 2 (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-33-10-301](https://user-images.githubusercontent.com/43462099/79043013-f345d000-7c2e-11ea-9da2-cfaf1d83537d.jpg)

I-War - Displays properly and no flickering or glitching
![I-War (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-30-55-382](https://user-images.githubusercontent.com/43462099/79043025-153f5280-7c2f-11ea-848f-f39db37583b2.jpg)
![I-War (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-31-29-703](https://user-images.githubusercontent.com/43462099/79043027-18d2d980-7c2f-11ea-811f-2f3fc1cd2300.jpg)

Kasumi Ninja - Displays properly, no missing textures/sprites
![Kasumi Ninja (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-34-02-268](https://user-images.githubusercontent.com/43462099/79043037-2d16d680-7c2f-11ea-8d6c-359711edff58.jpg)
![Kasumi Ninja (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-34-10-834](https://user-images.githubusercontent.com/43462099/79043039-3142f400-7c2f-11ea-9f27-e9ae89819e43.jpg)
![Kasumi Ninja (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-34-41-518](https://user-images.githubusercontent.com/43462099/79043043-356f1180-7c2f-11ea-9445-906ff781ef9e.jpg)

Missing Command 3D - The background textures aren't screwed and glitchy
![Missile Command 3D (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-35-18-999](https://user-images.githubusercontent.com/43462099/79043055-49b30e80-7c2f-11ea-9b3e-6495cbc4a2e6.jpg)

NBA Jam - Works fine without any glitches or jumping
![NBA Jam - Tournament Edition (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-36-27-469](https://user-images.githubusercontent.com/43462099/79043060-5899c100-7c2f-11ea-96f6-ea2ebf576de6.jpg)

Pinball Fantasies - Everything displays properly
![Pinball Fantasies (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-37-07-687](https://user-images.githubusercontent.com/43462099/79043072-69e2cd80-7c2f-11ea-8077-43749832ba04.jpg)

Power Drive Rally - Everything displays properly, and it doesn't crash.
![Power Drive Rally (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-38-19-454](https://user-images.githubusercontent.com/43462099/79043094-8bdc5000-7c2f-11ea-997a-d197a7247b56.jpg)

Ruiner Pinball - Works! Perfectly playable on Phoenix but doesn't work and is unsupported on Virtual Jaguar
![Ruiner Pinball (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-38-40-471](https://user-images.githubusercontent.com/43462099/79043106-ab737880-7c2f-11ea-9f94-92b914527401.jpg)
![Ruiner Pinball (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-38-54-888](https://user-images.githubusercontent.com/43462099/79043108-af06ff80-7c2f-11ea-8ff7-e4e123486399.jpg)
![Ruiner Pinball (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-39-09-627](https://user-images.githubusercontent.com/43462099/79043110-b201f000-7c2f-11ea-9cec-5bc9430fa88a.jpg)

Skyhammer - Works, and does NOT freeze like on Virtual Jaguar
![Skyhammer (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-40-16-347](https://user-images.githubusercontent.com/43462099/79043116-c6de8380-7c2f-11ea-940b-e3bfe922ade2.jpg)

Super Burnout - Displays properly, no flickering
![Super Burnout (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-40-48-656](https://user-images.githubusercontent.com/43462099/79043126-d9f15380-7c2f-11ea-8c50-907d0474032b.jpg)

Supercross 3D - Displays normally, I didn't realize the game textures were naturally like that for this game. This emulators shows that.
![SuperCross 3D (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-41-46-178](https://user-images.githubusercontent.com/43462099/79043137-f55c5e80-7c2f-11ea-91eb-0fc113051138.jpg)

Syndicate - No annoying pink glitches during menus
![Syndicate (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-42-20-757](https://user-images.githubusercontent.com/43462099/79043144-07d69800-7c30-11ea-8cd4-0673b98f567c.jpg)

Tempest 2000 - Works fine, with no flickering
![Tempest 2000 (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-43-04-386](https://user-images.githubusercontent.com/43462099/79043150-158c1d80-7c30-11ea-84ef-69fbbcc581ec.jpg)

Towers II - Works! Doesn't work on Virtual Jaguar
![Towers II (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-43-39-041](https://user-images.githubusercontent.com/43462099/79043176-33598280-7c30-11ea-8848-d51ac178b24e.jpg)
![Towers II (World)@ BIOS  Atari Jaguar (World)@Sat April 11 2020 11-44-01-691](https://user-images.githubusercontent.com/43462099/79043182-36ed0980-7c30-11ea-94ca-40640cca8349.jpg)

To be continued...

**Comment by JoeMatt** [comment]
I have some hacks, and I mean hacks, to improve some speed.

https://github.com/Provenance-Emu/virtualjaguar-libretro/tree/master

I haven't looked at this in 2 years, but I do remember some simple hacks such as using C struct's wiser for bit shift operands gave me a ton of speed up in the blitter memory copy and other functions, at least on ARM64.

For some reason, when profiling, some really basic bitwise ops were super slow. Using structs to represent the registers and casting using different C struct packings.

8f0f57375a922c86be28624d34891e992667a74e
2d5f483ae5483b0269aa20b36963212b8e5d2fae
8d25cab08133e8cd1c60fc9971d5c6f42db250b0

Inlining some things,
2d5f483ae5483b0269aa20b36963212b8e5d2fae
0285ac2b1f3eb824c13d94716d4105299dfffb83

Also some loop unrolling, 
c6bc3709c7177e25bd0a872a549f338d55f83d68

I'm using these hacks in Provenance. I haven't tested in forver, but users report pretty decent success with our Jaguar core and we don't even have options to switch the bios from internal or provided, which is needed for many games. I'm assuming these improvements have helped more than hurt but not sure if they'll be cross platform.

The struct packings should be endian agnostic based on using macros and the fact that I can test well in iOS simulator (intel) and device (arm) which have different endianeness.

YMMV on other platforms.

**Comment by JoeMatt** [comment]
I made a PR,
https://github.com/libretro/virtualjaguar-libretro/pulls

I fixed the CI issues and it even improved 3 CI warnings!
Maybe someone can build and run on retroach linux or windows to test?

I use a completely different front end so I don't have a build env setup on my Mac for retroarch atm.

**Comment by Cee123** [comment]
Wow, that's awesome! So going to try this, and test it out. That's awesome that someone has made progress on this. Fantastic work!

**Comment by Cee123** [comment]
I just tested it out on Odroid XU4 and my Odroid N2+ and couldn't get it to work on either of them. Each time I loaded a game, it would just freeze on a black screen.

Odroid XU4 - Tested with RetroArena/Ares, Retroarch 1.8.5
[virtualjaguar_libretro (Odroid XU4).zip](https://github.com/libretro/virtualjaguar-libretro/files/7174547/virtualjaguar_libretro.Odroid.XU4.zip)

Odroid N2 Plus - Tested with RetroArena/Ares and EmuElec
[virtualjaguar_libretro (Odroid N2+).zip](https://github.com/libretro/virtualjaguar-libretro/files/7174549/virtualjaguar_libretro.Odroid.N2%2B.zip)

![20210916_102544-20210916-102615506](https://user-images.githubusercontent.com/43462099/133540693-e1d61b7a-176e-4cf4-96c3-0e08e3a357a1.jpg)

![20210916_102459-20210916-10261620](https://user-images.githubusercontent.com/43462099/133540707-27d6ffc5-dde3-41ff-9821-13cbd3412070.jpg)

I also tried to compile it for Windows but got a bunch of error messages

![Capture](https://user-images.githubusercontent.com/43462099/133540944-85ab9040-e8f4-4067-80e5-e37bc3bb048c.PNG)

**Comment by JoeMatt** [comment]
Thanks for the reply.

The windows issues are probably because at the time I wasn't considered with backwards compatibility, was just trying to get something to work so I probably broke a bunch of cross platform things.

The the former issue, it does at least get to the bios screen?
To me that's important data because that means it's linking and running. It may be another platform issue past that though (probably path handling or something I made iOS/Mac specific, not remembering what I did two years ago).

That's promising though because those are easy fixes.

The core works on iOS/Mac for me so the code should be good, looks like a porting platform issue in that case.

**Comment by Cee123** [comment]
That's alright, no problem. Thank you for your work!

Yeah it gets to the bios screen but then freezes. Not sure if the bios needs to be turned on or off. I've tried with both. But I've noticed a MASSIVE difference in which the speed of the bios actually loads. It was slow before on both my XU4 and my N2. But now it's actually quicker. If the games themselves run like this they should be perfectly playable. This is interesting.

I might get some log files maybe so we can have some idea of what the issue might be. 

Yeah probably is a porting issue. If it works on iOS and Mac then that's good news. It should be able to get working on Windows, Linux (including arm and arm64) and Android.

**Comment by Cee123** [comment]
Unfortunately looks like the log files don't reveal much information. So no idea. Tested with the Odroid XU4

Alien Vs Predator
[retroarch__2021_09_06__11_48_28.log](https://github.com/libretro/virtualjaguar-libretro/files/7185984/retroarch__2021_09_06__11_48_28.log)
[retroarch__2021_09_06__11_49_20.log](https://github.com/libretro/virtualjaguar-libretro/files/7185985/retroarch__2021_09_06__11_49_20.log)

Checkered Flag
[retroarch__2021_09_06__11_50_07.log](https://github.com/libretro/virtualjaguar-libretro/files/7185989/retroarch__2021_09_06__11_50_07.log)
[retroarch__2021_09_06__11_51_00.log](https://github.com/libretro/virtualjaguar-libretro/files/7185994/retroarch__2021_09_06__11_51_00.log)

Doom
[retroarch__2021_09_06__11_51_49.log](https://github.com/libretro/virtualjaguar-libretro/files/7185997/retroarch__2021_09_06__11_51_49.log)
[retroarch__2021_09_06__11_52_26.log](https://github.com/libretro/virtualjaguar-libretro/files/7185998/retroarch__2021_09_06__11_52_26.log)

Raiden
[retroarch__2021_09_06__11_53_49.log](https://github.com/libretro/virtualjaguar-libretro/files/7186000/retroarch__2021_09_06__11_53_49.log)
[retroarch__2021_09_06__11_54_42.log](https://github.com/libretro/virtualjaguar-libretro/files/7186001/retroarch__2021_09_06__11_54_42.log)

Rayman
[retroarch__2021_09_06__11_55_45.log](https://github.com/libretro/virtualjaguar-libretro/files/7186003/retroarch__2021_09_06__11_55_45.log)
[retroarch__2021_09_06__11_56_37.log](https://github.com/libretro/virtualjaguar-libretro/files/7186005/retroarch__2021_09_06__11_56_37.log)

**Comment by Cee123** [comment]
I have to say, I was really impressed with the performance the last time I updated this emulator on my PC. I noticed the improvements straight away! It ran very good! Keep up the excellent work!

**Comment by Tasosgemah** [comment]
> 
> 
> I have to say, I was really impressed with the performance the last time I updated this emulator on my PC. I noticed the improvements straight away! It ran very good! Keep up the excellent work!

Are we still talking about the libretro core? Because that one is still 2.1.0 version.

**Comment by Cee123** [comment]
> 
> 
> > I have to say, I was really impressed with the performance the last time I updated this emulator on my PC. I noticed the improvements straight away! It ran very good! Keep up the excellent work!
> 
> Are we still talking about the libretro core? Because that one is still 2.1.0 version.

Yeah, it's still the same version. But he seems to have added some enhancements in performance. Something that's been long overdue.

**Comment by flytrapdev** [comment]
Apologies for the bump.

The emulator doesn't seem to account for changes in horizontal resolution in some of the games.
After taking a quick glance at the code, it seems like the horizontal resolution is hard-coded to 320 pixels.
This works for most games, however when a game changes this resolution, the emulator doesn't take it into account.
For instance in the case of Atari Karts, the in-game resolution is ~256x240, but the output resolution is still set to 320x240, which is why only part of the screen is used.

A "simple" fix would be to use the `TOMGetVideoModeWidth()` function from `src/tom.c` to detect changes in resolution during gameplay and update the av_info accordingly.
I tried it myself, but the best outcome I got was this :
![image](https://user-images.githubusercontent.com/68333938/202169180-ed356a02-3b93-487c-a318-21e3be5ec855.png)

The function doesn't return the "true" horizontal resolution, it returns a slightly larger value (326 instead of 320 etc.)
There is probably a way to determine the effective resolution but I'm not sure how, as I'm not familiar at all with the Jaguar hardware.

Maybe someone more familiar with the Jaguar could help...

I'm also not too familiar with the libretro functions (e.g `retro_get_system_av_info`), in the sense that I'm not sure what the different fields represent in relation to the video buffer (max_width, game_width etc.)

I'll try reading some of the Jaguar hardware docs to see if I can find something useful.
If anybody has more information on this please reach out !

**Comment by flytrapdev** [comment]
After some tinkering I found a potential solution.

Basically, the "screen pitch" needs to be updated when the resolution is changed.
With some hardcoded values I was able to get the correct aspect ratio in Atari Karts :

![image](https://user-images.githubusercontent.com/68333938/202288442-6d49877f-5833-4c34-baf0-01a450fe9d77.png)

I tried setting it automatically but that crashed the emulator, so there's still some work to do to make it work with other games.

I'm also not sure how this will affect Doom, since that game changes video modes during rendering.

**Comment by hizzlekizzle** [comment]
Ah, yeah, this is a common thing for cores to do. You can take a look at beetle-pce for an example how they handle it: https://github.com/libretro/beetle-pce-libretro/blob/master/libretro.cpp

**Comment by JoeMatt** [comment]
Oh wow! 

Nice stuff guys!

I'm super busy on Provenance, patching some other EMUs to iOS and a new project called SideStore.

I can def review and test anything but can't really write code at the moment.

Still plan to one day update the blitter...

**Comment by JoeMatt** [comment]
> I'm also not too familiar with the libretro functions (e.g `retro_get_system_av_info`), in the sense that I'm not sure what the different fields represent in relation to the video buffer (max_width, game_width etc.)

I'm pretty familiar with this. What you looking to do?

> I'm also not sure how this will affect Doom, since that game changes video modes during rendering.

Doom already has bespoke code so don't fret on that one for now. 
On Provenance we auto detect doom and turn on doom hacks, I think retroarch does too actually and also have args for it.

We could just use the doomResHack flag to skip stuff that's broken if needed.

**Comment by JoeMatt** [comment]
> Ah, yeah, this is a common thing for cores to do. You can take a look at beetle-pce for an example how they handle it: https://github.com/libretro/beetle-pce-libretro/blob/master/libretro.cpp

https://github.com/libretro/beetle-pce-libretro/blob/816b888bc33604df3fef0fb2ca309c0578ae52ac/libretro.cpp#L1650

**Comment by flytrapdev** [comment]
Thanks for the information !

I thought it was related to the libretro functions but apparently not, setting the screen pitch to the new resolution was enough. I'm still updating the av_info when the resolution changes however.
Although for some reason, when the new display width is higher than 512, the screen remains black. I'm not sure what's causing this, I thought it was a hard-coded limit in the emulator but apparently it's not.
This happens when the width is set to 652, as is the case in the Pinball Fantasies intro and the I-War and Aircars title screens.
I'm tempted to just halve the width and skip every other pixel when that happens.

I've also noted that some games crash Retroarch completely. Aircars crashes at the title screen, and Atari Karts crashes after playing for a while. I tried it with the regular virtual jaguar core and it also happened, so I'm guessing it's an unrelated issue.

**Comment by flytrapdev** [comment]
I found the fix for hi-res title screens : the libretro core max width was too low.

Here is the I-War title screen at the correct resolution.

![image](https://user-images.githubusercontent.com/68333938/202476545-ab19a0e3-d4d3-4524-a078-ffc4d4b5810a.png)

So far, this seems to fix resolution issues in the following games :
Atari Karts, Club Drive, Evolution Dino Dudes, Flashback, I-War, Pinball Fantasies and Trevor McFur.

Should I submit a pull request for this ?

**Comment by hizzlekizzle** [comment]
Yeah, sounds like a good improvement.

**Comment by Tasosgemah** [comment]
Thanks for the update. It's nice to see some game fix/updates for Virtual Jaguar after decades of being abandoned.

**Comment by TheDirtyWilly** [comment]
Excuse my ignorance here...  @flytrapdev you seem to have found a great fix for nearly all of the Jaguar games in this core!  How does one go about applying this fix to the core?  I saw you edited the .C file with the resolution changes, but does it need to be recompiled now?  And if so how do I obtain the latest version, or do we have to wait on the author?

**Comment by hizzlekizzle** [comment]
Yes, it needs to be recompiled with the fix. Updated builds should be available from RetroArch's online updater and/or buildbot.libretro.com in the nightly/.../latest directories.

**Comment by TheDirtyWilly** [comment]
Thanks!  To throw a monkey wrench in the works, I was hoping to get this working on the NVIDIA Shield core.  I've tried the online updater and there are no changes yet.  Still trying to get info on manually updating the Shield RetroArch cores.

**Comment by hizzlekizzle** [comment]
you can download it manually from the buildbot site and then use the "install a core" routine from the 'load core' menu to move it into the correct location.

However, you may want to just sideload the apk from the website, which pulls directly from the buildbot instead of using the Play Store servers, since Google won't allow us to update the packages on their Store for the foreseeable future.

**Comment by TheDirtyWilly** [comment]
TY @hizzlekizzle, this worked beautifully!  I used the buildbot.libretro.com nightly/android/latest/arm64-v8a/virtualjaguar_libretro_android.so.zip and the "Load Core" > "Install or Restore a Core" routine.  I may sideload the latest RetroArch.apk though, does seem like a better option.

**Comment by Cee123** [comment]
I'm impressed with the work done on this. Seems to work really well. You've done an awesome job. I've tested the games on this list with the cut off screens and the display to one side and now there are no issues at all! Not only that, but even the performance of this emulator is better! This is the most progress I've seen on this emulator in decades.

**Comment by Tasosgemah** [comment]
Depends, what do you mean resolved?

**Comment by JoeMatt** [comment]
Aleardy fixed a bunch of bugs in both HLE bios and with real bios.

Not all of these but many new fixes, Battle Sphere workingn with and without real bios, White Men Can't Jump plays with real bios, boots but fails to play with HLE and has jumpy graphics with both.

AvP map still messed up with "new" blitter and a few other bugs but tons of these I'm happy to report look fixed. Still going through the list.


#119 is the PR that will fix a lot of these.

**Comment by JoeMatt** [comment]
> I've decided to retract the bounty on this. It's been well over a year now and nothing's happened. My bounty of $250 has been refunded. During that time, unfortunately I saw no contributions or anyone volunteering to take on the bounty. Hopefully we do see some kind of continued development on a Jaguar emulator in the future but I'm not holding my breath anymore.

you can breathe easy now!

Still some bugs to fix but got a huge chunk of them, will breakout more PRs with specfiic fixes for the remainder, then onto CD and maybe even networking

**Comment by JoeMatt** [comment]
## v2.2.0 status (released as #119)

v2.2.0 is out — closing out a chunk of #38's sub-issues. Per-game status below; full table is also in [`docs/emulation-bug-hunt-todos.md`](https://github.com/libretro/virtualjaguar-libretro/blob/master/docs/emulation-bug-hunt-todos.md) under "Game compatibility (v2.2.0)".

### Fixed in v2.2.0

| Game | What changed |
|------|--------------|
| Air Cars | Title-screen cut-off |
| Atari Karts | Vertical bar artefact |
| Battlesphere Gold | Menu black-screen on real-BIOS path (fast-blitter menu glitch is a separate issue, see below) |
| Club Drive | Title cut-off, missing textures |
| Cybermorph | Missing/glitched textures; Skylar voice |
| Doom | Vertical bar; legacy "Doom resolution hack" core option removed (proper PWIDTH pixel replication) |
| Flashback | Display anchored to left |
| I-War | Title cut-off and gameplay glitches (accurate blitter) |
| Kasumi Ninja | Title flicker, missing textures, post-title freezes |
| Missile Command 3D | Minor flickering / glitches |
| Pinball Fantasies | Intro screens cut off |
| Powerdrive Rally | Intro cut-off + new-game crash |
| Skyhammer | Gameplay no longer freezes (audio is a separate, pre-existing issue) |
| Supercross 3D | Display anchored to left, blurry textures |
| Syndicate | Pink patches in text (accurate blitter) |
| Trevor McFur | Loading-screen cut-off |
| Val D'Isere | Display anchored to left |
| Wolfenstein 3D | HLE boot now works (geometry-pre-render fix; audio still broken — see below) |
| Zero 5 | Mild flickering |
| Zoop | Crash on launch |
| Raiden / Kasumi Ninja / etc. (HLE BIOS) | Many titles that previously needed a real BIOS image now boot on HLE |

### Still broken (carry into v2.3.0)

| Game | Symptom | Notes |
|------|---------|-------|
| Battle Sphere | Targeting reticle artefacts (accurate blitter) | Fast blitter renders correctly |
| Battlesphere Gold | Menu rendering glitches (fast blitter only) | |
| Fight for Life | Menu flickering | |
| Hover Strike | Crashes during gameplay | Fails identically with real BIOS — engine-level emulation bug, not HLE |
| Hyper Force | Black screen after title | Fails identically with real BIOS |
| Iron Soldier | Black screen after character select | Fails identically with real BIOS |
| Iron Soldier 2 | Black screen after character select; title music clipped | Audio clipping confirmed pre-existing on `libretro/master` (this PR didn't introduce it) |
| NBA Jam TE | Occasional screen jumping | Same engine pattern as White Men Can't Jump (both High Voltage) |
| Raiden | Won't boot on HLE | Real BIOS works. Cart's poll loop never sees its interrupt-handler-set flag (interrupts not arriving in HLE) |
| Ruiner Pinball | Won't boot in any config | Fails identically with real BIOS — `$402C`/`$4068` workspace pointers never get populated |
| Skyhammer / IS2 / Wolfenstein 3D audio | Silent (Wolf3D) or square-wave clipped (Skyhammer/IS2) | Pre-existing on `master`. Cause: cart-side DSP audio engine relies on BIOS-loaded DSP register-bank state we don't replicate in HLE. Documented with concrete next steps. |
| Super Burnout | Background-only render in gameplay | Fails identically with real BIOS — OP fixed-bitmap pipeline issue |
| Tempest 2000 | Rare flickering; accurate blitter slower than fast | |
| Towers II | Flickering/flashing during gameplay | PAL is worse; PAL audio broken |
| White Men Can't Jump | Heavy flickering, can't get past main menu on HLE | Real BIOS doesn't load splash either |

### Regressions introduced in this PR (also carry-forward)

| Game | Regression |
|------|------------|
| Val D'Isere | Player sprite z-order vs snow layer reversed (snow draws over skier) |
| I-War (fast blitter) | Title cut-off returns when fast blitter is enabled (accurate blitter unaffected) |
| I-War (accurate blitter) | Floor-layer flicker during gameplay |
| Kasumi Ninja (some stages) | Vertical tearing 3/4 down the screen |
| Syndicate (fast blitter) | Pink patches in text (accurate blitter is fine) |

### Cross-cutting finding

Of the 5 hang/crash titles in the still-broken list (Hyper Force, Iron Soldier, Hover Strike, Ruiner Pinball, Super Burnout), **4 fail identically with real BIOS and HLE BIOS** — same stuck PC, same pixel counts, same DSP state. They are real engine-level emulation bugs, not HLE-init issues. Real BIOS does not unblock them. Tracked individually in `docs/emulation-bug-hunt-todos.md`.

**Comment by Tasosgemah** [comment]
I would like to test these changes when the core updates in the buildbot.

**Comment by JoeMatt** [comment]
> I would like to test these changes when the core updates in the buildbot.

The release https://github.com/libretro/virtualjaguar-libretro/releases/tag/v2.2.0 should already have the update libs.

The included .info file needs to be overridden to force RA to load the full feature set.

**Comment by Tasosgemah** [comment]
DOOM runs way too fast with this, like it's running in fast forward speed

**Comment by JoeMatt** [comment]
> DOOM runs way too fast with this, like it's running in fast forward speed

Just noticed that too, must have been a later change.

Working on more fixes this weekend so may have another patch update for this soon.

**Comment by JoeMatt** [comment]
Housekeeping on a 2019 omnibus ticket.

This list of ~20 titles has been overtaken by six years of work, and its format now works against it: a single issue covering that many unrelated bugs cannot be closed, cannot be prioritised, and cannot record which entries are fixed. Several are known-fixed (the Iron Soldier 2 and Hover Strike entries were addressed during the Jaguar CD work; the Iron Soldier 3D-demo black screen no longer reproduces in any blitter × BIOS combination), several were never reproducible as described, and some are still live.

Per-title triage now lives in `docs/cart-issue-triage.md`, which records a verdict, the evidence behind it, and a copy-pasteable headless repro for each. Live items have their own tickets — #138 (Pitfall), #178 (Alien vs Predator), #187 (Tempest 2000).

Suggest **closing this** in favour of those. If a title on the original list is still broken for you on a [nightly](https://github.com/libretro/virtualjaguar-libretro/releases/tag/nightly), please open an issue for that title with a screenshot and the RetroArch log — the core now has a watchdog that names the failing subsystem (`gpu_pc_escape`, `video_stall`, `cd_seek_wedge`, …) at the moment it breaks, which makes a single-title report far more actionable than this list ever was.

**Comment by JoeMatt** [comment]
Thanks for the detailed 2019 report - it's been a useful checklist over the years. This emulator has had a lot of work since then (interrupt handling, bus/timing modeling, blitter accuracy, CD support, geometry negotiation), so it's worth a re-pass rather than leaving a seven-year-old list open indefinitely.

Several items from this list already have their own dedicated, previously-verified fixes and are tracked separately rather than re-litigated here:
- **Doom** ("weird bar down the screen") - speed/timing calibrated against ground truth, see the Doom pace calibration docs and #265.
- **Iron Soldier / Iron Soldier 2** black screens - resolved (match-load hang + cold-boot black screen, both root-caused and fixed).
- **Hover Strike** (glitched/flickering) - resolved (streamed CD_read fix).
- **Pitfall** (`#138`, black screen after ~1 min) - resolved (68000 group-0 frame + GPU STOREP masking fix).

For the rest, I did a smoke-level headless re-verification pass on current `develop` of the titles from this list that had never been individually re-checked: Air Cars, Atari Karts, Club Drive, Cybermorph, Fight For Your Life (listed as "Fight for Life"), Flashback, Hyper Force, and I-War. Method: 2400 frames headless per title, no scripted input beyond holding A from frame 300, checking (a) the core boots to a non-black, changing framebuffer, (b) the runtime crash-detect watchdog (`gpu_pc_escape`, `dsp_pc_escape`, `gpu_wedge`, `dsp_wedge`, `video_stall`) for hard failures, (c) audio presence via the harness's non-silent sample count, plus one screenshot per title. This is a **smoke test, not a deep functional re-verification** - it can't confirm subtle in-game rendering artifacts, only that a title isn't crashed, wedged, black-screened, or silent, and directly rebuts a handful of the original screen-geometry claims where the captured frame was the exact screen described.

| Title | 2019 claim | Current status | Evidence |
|---|---|---|---|
| Air Cars | Half the title screen is cut off | **Not reproduced.** Full, uncropped "Game Selection" menu renders | Screenshot at the described screen; no crash signature; audio present (1.22M/1.92M non-silent samples) over 2400 frames |
| Atari Karts | Weird bar down the side during the game | **Not adjudicated in-game.** Boots clean, no crash signature, actively animating (31% frame-to-frame change) | Only the title card was captured this pass, not gameplay, so the specific in-game bar wasn't directly checked; no crash/black-screen either way |
| Club Drive | Half title screen cut off + missing textures | **Not reproduced.** Full title scene (palm trees, car, castle) renders with textures intact | Screenshot at the described screen; no crash signature; audio 98% non-silent |
| Cybermorph | Missing/glitched textures | **Not reproduced** at the gameplay HUD reached this pass - all HUD elements render | Screenshot in gameplay; no crash signature; audio 98% non-silent |
| Fight For Your Life | Flickering graphics, mainly in the menu | **Not reproduced** in the captured frame; menu renders cleanly | Screenshot of menu; zero crash-detect signatures across the full 2400 frames (the only title of the eight with none); audio 91% non-silent |
| Flashback | Game displays on the left side of the screen | **Not adjudicated in-game.** Publisher splash renders centered, not left-shifted | Only the splash screen was captured this pass, not gameplay; no crash signature; audio 91% non-silent |
| Hyper Force | Black screen after title, doesn't work at all | **Not verified** - ROM not present in this test collection | Skipped; no evidence either way |
| I-War | Half title screen cut off, glitched/flickering | **Not reproduced.** Full-width title card renders | Screenshot at the described screen; no crash signature; audio 97% non-silent |

None of the seven checked titles showed a crash-detect signature, a permanent black screen, or silent audio. Where the original claim was specifically about a title/menu screen (Air Cars, Club Drive, I-War) or an in-gameplay HUD (Cybermorph), the captured frame is a direct rebuttal. Where the claim was about active gameplay we didn't drive into (Atari Karts' in-game bar, Flashback's in-game framing), this pass only confirms "boots clean, no crash" and doesn't settle the specific visual claim - genuine re-verification of those would need scripted input into gameplay plus a rendering comparison, which is out of scope for a smoke sweep.

Given the age of this report, the number of unrelated fixes it's accumulated over the years, and that nothing in this pass turned up a reproducible crash or black-screen to file as a new issue, I'm closing this as superseded. If Atari Karts' in-game bar or Flashback's in-game left-shift are still present, please open a fresh, single-title issue with a screenshot from actual gameplay - that's far easier to act on than a resurrected seven-year-old laundry list.

**Comment by JoeMatt** [comment]
Closing per the re-verification above.

**Comment by JoeMatt** [comment]
Thanks for the report, and for the follow-ups on it.

A correction on the Doom item: its pace is not fixed in v3.0.0. What shipped
is the measurement — we calibrated our render rate against two independent
real-hardware captures and can now state the gap precisely: silicon runs the
engine at 19.98 fps typical (~15 fps in heavy scenes) while we sit pinned at
the engine's own 29.97 fps two-field cap, so combat and monster pacing run
~1.5x fast typically and ~2x in heavy scenes. The write-up, including why the
tic-rate metric everyone (us included) was using is saturated and can't detect
this, is in `docs/doom-pace-calibration.md`. It's listed under Known Issues in
the v3.0.0 notes, and the sensitivity sweeps point at GPU/DSP instruction
timing as the actual lever — that's the next piece of work, not something
v3.0.0 closes.

On the "weird bar down half the screen" artifact from the original report: we
haven't been able to re-verify that one, and a lot has changed in the video
path since. If it still shows up on v3.0.0, a fresh issue with a screenshot
and the ROM/settings would be the most useful thing — it'll get triaged on its
own rather than as part of this thread.

One disambiguation that's caused some confusion: Iron Soldier (cart) and Iron
Soldier 2 (CD) are separate titles here. The Iron Soldier items resolved in
v3.0.0 are IS2 CD bugs — the match-load hang and the cold-boot black screen.
Iron Soldier 1's missing wireframe tank under the fast blitter is still open
and is listed in the v3.0.0 Known Issues; the accurate blitter is unaffected,
so that's the workaround for now.
